torch/nn/utils/parametrizations.py

```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
     return module
 
 
+class _WeightNorm(Module):
+    def __init__(
+        self,
+        dim: Optional[int] = 0,
+    ) -> None:
+        super().__init__()
+        if dim is None:
+            dim = -1
+        self.dim = dim
+
+    def forward(self, weight_g, weight_v):
+        return torch._weight_norm(weight_v, weight_g, self.dim)
+
+    def right_inverse(self, weight):
+        weight_g = torch.norm_except_dim(weight, 2, self.dim)
+        weight_v = weight
+
+        return weight_g, weight_v
+
+
+def weight_norm(module: Module, name: str = 'weight', dim: int = 0):
+    r"""Applies weight normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{w} = g \dfrac{\mathbf{v}}{\|\mathbf{v}\|}
+
+    Weight normalization is a reparameterization that decouples the magnitude
+    of a weight tensor from its direction. This replaces the parameter specified
+    by :attr:`name` with two parameters: one specifying the magnitude
+    and one specifying the direction.
+
+    By default, with ``dim=0``, the norm is computed independently per output
+    channel/plane. To compute a norm over the entire weight tensor, use
+    ``dim=None``.
+
+    See https://arxiv.org/abs/1602.07868
+
+    Args:
+        module (Module): containing module
+        name (str, optional): name of weight parameter
+        dim (int, optional): dimension over which to compute the norm
+
+    Returns:
+        The original module with the weight norm hook
+
+    Example::
+
+        >>> m = weight_norm(nn.Linear(20, 40), name='weight')
+        >>> m
+        ParametrizedLinear(
+          in_features=20, out_features=40, bias=True
+          (parametrizations): ModuleDict(
+            (weight): ParametrizationList(
+              (0): _WeightNorm()
+            )
+          )
+        )
+        >>> m.parametrizations.weight.original0.size()
+        torch.Size([40, 1])
+        >>> m.parametrizations.weight.original1.size()
+        torch.Size([40, 20])
+
+    """
+    _weight_norm = _WeightNorm(dim)
+    parametrize.register_parametrization(module, name, _weight_norm, unsafe=True)
+
+    def _weight_norm_compat_hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
```

Comment:
also, should installation of this hook be optional? (to avoid surprising interaction / influence with other hooks)

Reply:
And maybe still remove the underscore from _WeightNorm?

Reply:
@lezcano can you explain to us why the preexisting parametrizations are underscore'd on their module?

Reply:
The idea is that they are not exposed as public API (that API is not documented) hence the underscore.

Reply:
I guess the question is why are they not exposed... some time has passed since their introduction, so maybe now could make them public api

Reply:
To expose more the parameterization design to general public (so that people would take the WeightNorm code as code example - and I think it's mature enough for this), signal that it's mature, to encourage less use of "magical" methods `register_weight_norm` and expose users to compositionality of parametrizations.

If these are public, the current `register_weight_norm` and its friends can be deprecated.

Reply:
I must admit, I do find making the Parametrization modules private a bit odd. We don't make normal nn Modules private; why should parametrizations be different? (That being said I admit the precautionary principle says that you should opt to make things private if there is no reason for them to be public.)

Reply:
Yes, but I propose that people directly use `parametrize.register_parametrization` for that. And deprecating the special method `weight_norm(...)` and asking to directly use `parametrize.register_parametrization` would highlight this compositionality - especially if this helper now does nothing except create a `_WeightNorm` and call `parametrize.register_parametrization`

Reply:
I think in earlier days making them private was more justified because it was the move to the new parameterization interface (and the old way was calling a helper method), but now it's more stable and it makes sense to promote the direct use of parametrization modules like `_WeightNorm`...

Reply:
I do prefer these helper methods. I did something similar in my library https://github.com/Lezcano/geotorch/, and I very much prefer the functional API than the one where you use `register_parametrization` directly. But yeah, this is very subjective indeed.

Reply:
Btw `register_parametrization` seems to return `self`, so could be used in quasi-functional way as well! :)